linkerd/stack/src/queue.rs

```diff
@@ -1,93 +1,153 @@
+#[derive(Debug)]
+pub enum WithTimeout {}
+
+pub type NewQueueWithTimeout<X, Req, N> = NewQueue<WithTimeout, X, Req, N>;
+
+pub type QueueWithTimeout<Req, Rsp, E = Error> =
+    failfast::Gate<Buffer<BoxService<Req, Rsp, E>, Req>>;
+
+/// A marker type to indicate a queue does not apply a timeout.
+#[derive(Debug)]
+pub enum WithoutTimeout {}
+
+pub type NewQueueWithoutTimeout<X, Req, N> = NewQueue<WithoutTimeout, X, Req, N>;
+
+pub type QueueWithoutTimeout<Req, Rsp, E = Error> = Buffer<BoxService<Req, Rsp, E>, Req>;
 
 // === impl NewQueue ===
 
-impl<T, X, Req, N> NewService<T> for NewQueue<X, Req, N>
+impl<T, X, Req, N, S> NewService<T> for NewQueueWithTimeout<X, Req, N>
 where
     Req: Send + 'static,
-    X: ExtractParam<QueueConfig, T>,
-    N: NewService<T>,
-    N::Service: Service<Req> + Send + 'static,
-    <N::Service as Service<Req>>::Future: Send,
-    <N::Service as Service<Req>>::Error: Into<Error>,
-    <N::Service as Service<Req>>::Response: 'static,
+    X: ExtractParam<Capacity, T>,
+    X: ExtractParam<Timeout, T>,
+    N: NewService<T, Service = S>,
+    S: Service<Req> + Send + 'static,
+    S::Response: 'static,
+    S::Error: Into<Error>,
+    S::Future: Send,
 {
-    type Service = Queue<Req, <N::Service as Service<Req>>::Response>;
+    type Service = QueueWithTimeout<Req, S::Response>;
 
     fn new_service(&self, target: T) -> Self::Service {
-        let QueueConfig {
-            capacity,
-            failfast_timeout,
-        } = self.extract.extract_param(&target);
+        let Capacity(capacity) = self.extract.extract_param(&target);
+        let Timeout(failfast_timeout) = self.extract.extract_param(&target);
         let buf = layer::mk(move |inner| Buffer::new(BoxService::new(inner), capacity));
         let buf = FailFast::layer_gated(failfast_timeout, buf);
         buf.layer(self.inner.new_service(target))
     }
 }
 
-impl<X: Clone, Req, N> NewQueue<X, Req, N> {
+impl<T, X, Req, N, S> NewService<T> for NewQueueWithoutTimeout<X, Req, N>
+where
+    Req: Send + 'static,
+    X: ExtractParam<Capacity, T>,
+    N: NewService<T, Service = S>,
+    S: Service<Req> + Send + 'static,
+    S::Response: 'static,
+    S::Error: Into<Error> + Send + Sync + 'static,
+    S::Future: Send,
+{
+    type Service = QueueWithoutTimeout<Req, S::Response, S::Error>;
+
+    fn new_service(&self, target: T) -> Self::Service {
+        let Capacity(capacity) = self.extract.extract_param(&target);
+        let inner = self.inner.new_service(target);
+        Buffer::new(BoxService::new(inner), capacity)
+    }
+}
+
+impl<Z, X: Clone, Req, N> NewQueue<Z, X, Req, N> {
+    pub fn new(inner: N, extract: X) -> Self {
+        Self {
+            inner,
+            extract,
+            _req: PhantomData,
+        }
+    }
+
     /// Returns a [`Layer`] that constructs new [`failfast::Gate`]d [`Buffer`]s
     /// using an `X`-typed [`ExtractParam`] implementation to extract
     /// [`QueueConfig`] from a `T`-typed target.
```

Comment:
this comment is no longer correct, as this layer no longer constructs buffers with the failfast gate.